networking_cisco/plugins/cisco/db/l3/l3_router_appliance_gbp_db.py

```diff
@@ -0,0 +1,85 @@
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from networking_cisco.plugins.cisco.db.l3 import (
+    l3_router_appliance_db as l3_r_db)
+from networking_cisco.plugins.cisco.l3.drivers import driver_context
+
+from neutron.common import constants as l3_constants
+from neutron.common import exceptions as n_exc
+from neutron.i18n import _LW
+
+from oslo_config import cfg
+from oslo_log import log as logging
+
+FLOATINGIP_STATUS_ACTIVE = l3_constants.FLOATINGIP_STATUS_ACTIVE
+
+LOG = logging.getLogger(__name__)
+
+
+class L3RouterApplianceGbpDBMixin(l3_r_db.L3RouterApplianceDBMixin):
+    """Mixin class for Neutron's routing service using appliances with ACI.
+
+    This class should only be used for integrations with ACI using the
+    Group Based Policy (GBP) workflow. It's only difference from the parent
+    class is to specialize the create_floatingip call to support multiple
+    pools implemented by the GBP workflow.
+    """
+
+    def create_floatingip(self, context, floatingip,
+                          initial_status=FLOATINGIP_STATUS_ACTIVE):
+        """Specialization of parent class's create_floatingip.
+
+        This specialization of the parent class's create_floatingip
+        is needed for GBP workflow, as the pre-/post-commmit calls for
+        creating the floating IP must be peformed in a loop with the database
+        call in the grandparent class.
+        """
+
+        res = None
+        fip = floatingip['floatingip']
+        if not fip.get('subnet_id'):
+            # NOTE: default router type must be ASR1k
+            router_type_name = cfg.CONF.routing.default_router_type
+            driver = self._get_router_type_driver(context,
+                                                  router_type_name)
+            if driver:
+                fip_ctxt = driver_context.FloatingipContext(floatingip)
+                driver.create_floatingip_precommit(context, fip_ctxt)
+                nat_pool_list = getattr(context, 'nat_pool_list', [])
+                for nat_pool in nat_pool_list:
+                    if not nat_pool:
+                        continue
+                    fip['subnet_id'] = nat_pool['subnet_id']
+                    try:
+                        res = super(l3_r_db.L3RouterApplianceDBMixin,
+                                    self).create_floatingip(context,
+                                                            floatingip)
+                        router_ids = ([res['router_id']]
+                                      if res['router_id'] else [])
+                    except n_exc.IpAddressGenerationFailure as ex:
+                        LOG.warning(_LW("Floating allocation failed: %s"),
+                                    ex.message)
+                    if res:
+                        break
+        if not res:
+            info = super(l3_r_db.L3RouterApplianceDBMixin,
```

Comment:
if the create_floatingip() call failed at line#66 then this will get called again. However its basically the same call so it will fail again....is this what we intend to do?